datastore/redis.go

```diff
@@ -451,31 +465,34 @@ type SaveBidAndUpdateTopBidResponse struct {
 	WasTopBidUpdated bool // Whether the top bid was updated
 	IsNewTopBid      bool // Whether the submitted bid became the new top bid
 
-	TopBidValue   *big.Int
-	TopBidBuilder string
-
-	PrevTopBidValue   *big.Int
-	PrevTopBidBuilder string
+	TopBidValue     *big.Int
+	PrevTopBidValue *big.Int
 }
 
-func (r *RedisCache) SaveBidAndUpdateTopBid(payload *common.BuilderSubmitBlockRequest, getPayloadResponse *common.GetPayloadResponse, getHeaderResponse *common.GetHeaderResponse, reqReceivedAt time.Time, isCancellationEnabled bool) (state SaveBidAndUpdateTopBidResponse, err error) {
+func (r *RedisCache) SaveBidAndUpdateTopBid(payload *common.BuilderSubmitBlockRequest, getPayloadResponse *common.GetPayloadResponse, getHeaderResponse *common.GetHeaderResponse, reqReceivedAt time.Time, isCancellationEnabled bool, floorValue *big.Int) (state SaveBidAndUpdateTopBidResponse, err error) {
 	// 1. Load latest bids for a given slot+parent+proposer
 	keyBidValues := r.keyBlockBuilderLatestBidsValue(payload.Slot(), payload.ParentHash(), payload.ProposerPubkey())
 	bidValueMap, err := r.client.HGetAll(context.Background(), keyBidValues).Result()
 	if err != nil {
 		return state, err
 	}
 
+	if floorValue == nil {
+		floorValue, err = r.GetFloorBidValue(payload.Slot(), payload.ParentHash(), payload.ProposerPubkey())
+		if err != nil {
+			return state, err
+		}
+	}
+
 	builderBids := NewBuilderBids(bidValueMap)
-	state.PrevTopBidBuilder, state.PrevTopBidValue = builderBids.getTopBid()
-	state.TopBidBuilder, state.TopBidValue = state.PrevTopBidBuilder, state.PrevTopBidValue
+	_, state.PrevTopBidValue = builderBids.getTopBid()
+	state.TopBidValue = state.PrevTopBidValue
 
 	// 2. Do we even need to continue / save the new payload and update the top bid?
 	// - In cancellation mode: always continue to saving latest bid
-	// - In non-cancellation mode: only save if current bid is higher value than this builders previous bid
+	// - In non-cancellation mode: only save if current bid is higher value than floor value
 	if !isCancellationEnabled {
-		currentBuilderLastValue := builderBids.builderValue(payload.BuilderPubkey().String())
-		if payload.Value().Cmp(currentBuilderLastValue) < 1 {
+		if payload.Value().Cmp(floorValue) < 1 {
```

Comment:
can we just `&&` these in the same if clause ?

Reply:
yeah